File: bridge/polyfill/src/hybrid-history.ts
/*
* Copyright (C) 2019-2022 The Kraken authors. All rights reserved.
* Copyright (C) 2022-present The WebF authors. All rights reserved.
*/

import {webf} from './webf';

/**
 * Interface for managing navigation history in hybrid applications
 * Provides both web-like history API and Flutter-like navigation methods
 */
export interface HybridHistoryInterface {
  // State management
  /** Current state object associated with the history entry */
  readonly state: any;
  /** Current path of the history entry */
  readonly path: string;

  // Original API methods
  /**
   * Navigate back to the previous history entry
   */
  back(): void;

  /**
   * Push a new state to the history stack
   * @param state State object to associate with the new history entry
   * @param name Route name for the new history entry
   */
  pushState(state: any, name: string): void;

  /**
   * Replace the current history entry with a new one
   * @param state State object to associate with the new history entry
   * @param name Route name for the new history entry
   */
  replaceState(state: any, name: string): void;

  /**
   * Pop the current route and push a new one, with ability to restore
   * @param state State object to associate with the new history entry
   * @param name Route name for the new history entry
   * @returns A restoration identifier that can be used to restore this state later
   */
  restorablePopAndPushState(state: any, name: string): string;

  // Flutter-like API methods
  /**
   * Close the current screen and return to the previous one
   * @param result Optional result to pass back to the previous screen
   */
  pop(result?: any): void;

  /**
   * Navigate to a named route
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   */
  pushNamed(routeName: string, options?: { arguments?: any }): void;

  /**
   * Replace the current route with a new named route
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   */
  pushReplacementNamed(routeName: string, options?: { arguments?: any }): void;

  /**
   * Pop the current route and push a new one, with ability to restore
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   * @returns A restoration identifier that can be used to restore this state later
   */
  restorablePopAndPushNamed(routeName: string, options?: { arguments?: any }): string;

  // Utility methods
  /**
   * Check if the navigator can go back
   * @returns True if there are routes that can be popped, false otherwise
   */
  canPop(): boolean;

  /**
   * Pop the current route if possible
   * @param result Optional result to pass back to the previous screen
   * @returns True if the route was popped, false otherwise
   */
  maybePop(result?: any): boolean;

  /**
   * Pop the current route and push a new named route
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   */
  popAndPushNamed(routeName: string, options?: { arguments?: any }): void;

  /**
   * Pop routes until reaching a specific route
   * @param routeName Name of the route to stop at
   */
  popUntil(routeName: string): void;

  /**
   * Push a new route and remove routes until reaching a specific route
   * @param state State object to associate with the new route
   * @param newRouteName Name of the new route to navigate to
   * @param untilRouteName Name of the route to stop removing at
   */
  pushNamedAndRemoveUntil(state: any, newRouteName: string, untilRouteName: string): void;

  /**
   * Flutter-style method to push a new route and remove routes until reaching a specific route
   * @param newRouteName Name of the new route to navigate to
   * @param untilRouteName Name of the route to stop removing at
   * @param options Additional options like arguments to pass to the route
   */
  pushNamedAndRemoveUntilRoute(newRouteName: string, untilRouteName: string, options?: { arguments?: any }): void;
}

/**
 * Implementation of HybridHistory providing both web-like and Flutter-like navigation APIs
 */
class HybridHistory implements HybridHistoryInterface {
  // State management
  /**
   * Get the current state object associated with the history entry
   * @returns Current state as a parsed JSON object
   */
  get state() {
    return JSON.parse(webf.invokeModule('HybridHistory', 'state'));
  }

  /**
   * Get the current path of the history entry
   * @returns Current path as a string
   */
  get path() {
    return webf.invokeModule('HybridHistory', 'path');
  }

  // Original API methods - kept for backward compatibility
  /**
   * Navigate back to the previous history entry
   */
  back() {
    webf.invokeModule('HybridHistory', 'back');
  }

  /**
   * Push a new state to the history stack
   * @param state State object to associate with the new history entry
   * @param name Route name for the new history entry
   * @throws TypeError if arguments are missing
   */
  pushState(state: any, name: string) {
    if (arguments.length < 2) {
      throw TypeError("Failed to execute 'pushState' on 'HybridHistory': 2 arguments required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'pushState', state, name);
  }

  /**
   * Replace the current history entry with a new one
   * @param state State object to associate with the new history entry
   * @param name Route name for the new history entry
   * @throws TypeError if arguments are missing
   */
  replaceState(state: any, name: string) {
    if (arguments.length < 2) {
      throw TypeError("Failed to execute 'replaceState' on 'HybridHistory': 2 arguments required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'replaceState', state, name);
  }

  /**
   * Pop the current route and push a new one, with ability to restore
   * @param state State object to associate with the new history entry
   * @param name Route name for the new history entry
   * @returns A restoration identifier that can be used to restore this state later
   * @throws TypeError if arguments are missing
   */
  restorablePopAndPushState(state: any, name: string): string {
    if (arguments.length < 2) {
      throw TypeError("Failed to execute 'restorablePopAndPushState' on 'HybridHistory': 2 arguments required, but only " + arguments.length + " present");
    }
    return webf.invokeModule('HybridHistory', 'restorablePopAndPushState', state, name);
  }

  // New Flutter-like API methods
  /**
   * Close the current screen and return to the previous one
   * @param result Optional result to pass back to the previous screen
   */
  pop(result?: any) {
    webf.invokeModule('HybridHistory', 'pop', result !== undefined ? result : null);
  }

  /**
   * Navigate to a named route
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   * @throws TypeError if arguments are missing
   */
  pushNamed(routeName: string, options: { arguments?: any } = {}) {
    if (arguments.length < 1) {
      throw TypeError("Failed to execute 'pushNamed' on 'HybridHistory': 1 argument required, but only " + arguments.length + " present");
    }
    if (options.arguments !== undefined) {
      webf.invokeModule('HybridHistory', 'pushNamed', routeName, options.arguments);
    } else {
      webf.invokeModule('HybridHistory', 'pushNamed', routeName);
    }
  }

  /**
   * Replace the current route with a new named route
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   * @throws TypeError if arguments are missing
   */
  pushReplacementNamed(routeName: string, options: { arguments?: any } = {}) {
    if (arguments.length < 1) {
      throw TypeError("Failed to execute 'pushReplacementNamed' on 'HybridHistory': 1 argument required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'pushReplacementNamed', routeName, options.arguments !== undefined ? options.arguments : null);
  }

  /**
   * Pop the current route and push a new one, with ability to restore
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   * @returns A restoration identifier that can be used to restore this state later
   * @throws TypeError if arguments are missing
   */
  restorablePopAndPushNamed(routeName: string, options: { arguments?: any } = {}): string {
    if (arguments.length < 1) {
      throw TypeError("Failed to execute 'restorablePopAndPushNamed' on 'HybridHistory': 1 argument required, but only " + arguments.length + " present");
    }
    return webf.invokeModule('HybridHistory', 'restorablePopAndPushNamed', routeName, options.arguments !== undefined ? options.arguments : null);
  }

  // Utility methods
  /**
   * Check if the navigator can go back
   * @returns True if there are routes that can be popped, false otherwise
   */
  canPop(): boolean {
    return webf.invokeModule('HybridHistory', 'canPop') === 'true';
  }

  /**
   * Pop the current route if possible
   * @param result Optional result to pass back to the previous screen
   * @returns True if the route was popped, false otherwise
   */
  maybePop(result?: any): boolean {
    return webf.invokeModule('HybridHistory', 'maybePop', result !== undefined ? result : null) === 'true';
  }

  /**
   * Pop the current route and push a new named route
   * @param routeName Name of the route to navigate to
   * @param options Additional options like arguments to pass to the route
   * @throws TypeError if arguments are missing
   */
  popAndPushNamed(routeName: string, options: { arguments?: any } = {}) {
    if (arguments.length < 1) {
      throw TypeError("Failed to execute 'popAndPushNamed' on 'HybridHistory': 1 argument required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'popAndPushNamed', routeName, options.arguments !== undefined ? options.arguments : null);
  }

  /**
   * Pop routes until reaching a specific route
   * @param routeName Name of the route to stop at
   * @throws TypeError if arguments are missing
   */
  popUntil(routeName: string) {
    if (arguments.length < 1) {
      throw TypeError("Failed to execute 'popUntil' on 'HybridHistory': 1 argument required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'popUntil', routeName);
  }

  /**
   * Push a new route and remove routes until reaching a specific route
   * @param state State object to associate with the new route
   * @param newRouteName Name of the new route to navigate to
   * @param untilRouteName Name of the route to stop removing at
   * @throws TypeError if arguments are missing
   */
  pushNamedAndRemoveUntil(state: any, newRouteName: string, untilRouteName: string) {
    if (arguments.length < 3) {
      throw TypeError("Failed to execute 'pushNamedAndRemoveUntil' on 'HybridHistory': 3 arguments required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'pushNamedAndRemoveUntil', state, newRouteName, untilRouteName);
  }

  /**
   * Flutter-style method to push a new route and remove routes until reaching a specific route
   * @param newRouteName Name of the new route to navigate to
   * @param untilRouteName Name of the route to stop removing at
   * @param options Additional options like arguments to pass to the route
   * @throws TypeError if arguments are missing
   */
  pushNamedAndRemoveUntilRoute(newRouteName: string, untilRouteName: string, options: { arguments?: any } = {}) {
    if (arguments.length < 2) {
      throw TypeError("Failed to execute 'pushNamedAndRemoveUntilRoute' on 'HybridHistory': 2 arguments required, but only " + arguments.length + " present");
    }
    webf.invokeModule('HybridHistory', 'pushNamedAndRemoveUntil',
      newRouteName, untilRouteName, options.arguments !== undefined ? options.arguments : null);
  }
}

/**
 * Global instance of HybridHistory for managing navigation
 */
export const hybridHistory = new HybridHistory();

File: webf/lib/src/module/hybrid_history.dart
/*
 * Copyright (C) 2019-2022 The Kraken authors. All rights reserved.
 * Copyright (C) 2022-present The WebF authors. All rights reserved.
 */

import 'dart:convert';

import 'package:flutter/widgets.dart';
import 'package:webf/webf.dart';

class HybridHistoryItem {
  HybridHistoryItem(this.bundle, this.state, this.needJump);

  final WebFBundle bundle;
  final dynamic state;
  final bool needJump;
}

abstract class HybridHistoryDelegate {
  /// Pop the top-most route off the navigator that most tightly encloses the
  /// given context.
  void pop(BuildContext context);

  /// Push a named route onto the navigator that most tightly encloses the given
  /// context.
  void pushNamed(
    BuildContext context,
    String routeName, {
    Object? arguments,
  });

  String restorablePopAndPushNamed<T extends Object?, TO extends Object?>(
      BuildContext context,
      String routeName, {
        TO? result,
        Object? arguments,
      });

  void replaceState(BuildContext context, Object? state, String name);

  String path(BuildContext context, String? initialRoute);

  dynamic state(BuildContext context, Map<String, dynamic>? initialState);

  /// Pops until a route with the given name.
  void popUntil(BuildContext context, RoutePredicate predicate);

  /// Whether the navigator can be popped.
  bool canPop(BuildContext context);

  /// Pop the top-most route off, only if it's not the last route.
  Future<bool> maybePop<T extends Object?>(BuildContext context, [T? result]);

  /// Pop the current route off and push a named route in its place.
  void popAndPushNamed(
    BuildContext context,
    String routeName, {
    Object? arguments,
  });

  /// Push the route with the given name and remove routes until the predicate returns true.
  void pushNamedAndRemoveUntil(
    BuildContext context,
    String newRouteName,
    RoutePredicate predicate, {
    Object? arguments,
  });
}

class HybridHistoryModule extends BaseModule {
  @override
  String get name => 'HybridHistory';

  HybridHistoryDelegate? _delegate;

  set delegate(HybridHistoryDelegate value) {
    _delegate = value;
  }

  HybridHistoryModule(ModuleManager? moduleManager) : super(moduleManager);

  BuildContext? get _context => moduleManager?.controller.currentBuildContext?.context;

  void back() async {
    pop();
  }

  /// Pop the current route - matches Flutter's Navigator.pop
  void pop([Object? result]) {
    if (_delegate != null) {
      _delegate!.pop(_context!);
      return;
    }
    Navigator.pop(_context!, result);
  }

  /// Push a named route onto the navigator that most tightly encloses the given
  /// context.
  /// Original API - kept for backward compatibility
  void pushState(Object? state, String name) {
    pushNamed(name, arguments: state);
  }

  /// Push a named route onto the navigator - matches Flutter's Navigator.pushNamed
  void pushNamed(String routeName, {Object? arguments}) {
    if (_delegate != null) {
      _delegate!.pushNamed(_context!, routeName, arguments: arguments);
      return;
    }
    Navigator.pushNamed(_context!, routeName, arguments: arguments);
  }

  /// Pop the current route off the navigator that most tightly encloses the
  /// given context and push a named route in its place.
  /// Original API - kept for backward compatibility
  String restorablePopAndPushState(Object? state, String name) {
    return restorablePopAndPushNamed(name, arguments: state);
  }

  /// Restorably pop the current route and push a named route - matches Flutter's Navigator
  String restorablePopAndPushNamed(String routeName, {Object? arguments}) {
    if (_delegate != null) {
      return _delegate!.restorablePopAndPushNamed(_context!, routeName, arguments: arguments);
    }
    return Navigator.restorablePopAndPushNamed(_context!, routeName, arguments: arguments);
  }

  /// Original API - kept for backward compatibility
  void replaceState(Object? state, String name) {
    if (moduleManager!.controller.buildContextStack.length == 1) {
      throw FlutterError('Can not replaceState at the top stack of hybrid router');
    }
    pushReplacementNamed(name, arguments: state);
  }

  /// Push a replacement named route - matches Flutter's Navigator.pushReplacementNamed
  void pushReplacementNamed(String routeName, {Object? arguments}) {
    if (_delegate != null) {
      _delegate!.replaceState(_context!, arguments, routeName);
      return;
    }
    Navigator.pushReplacementNamed(_context!, routeName, arguments: arguments);
  }

  String path() {
    String initialRoute = moduleManager!.controller.initialRoute;
    if (_delegate != null) {
      return _delegate!.path(_context!, initialRoute);
    }
    String? currentPath = ModalRoute.of(_context!)?.settings.name;
    return currentPath ?? initialRoute ?? '/';
  }

  /// Whether the navigator can be popped.
  bool canPop() {
    if (_delegate != null) {
      return _delegate!.canPop(_context!);
    }
    return Navigator.canPop(_context!);
  }

  /// Pop the top-most route off, only if it's not the last route.
  Future<bool> maybePop([Object? result]) async {
    if (_delegate != null) {
      return await _delegate!.maybePop(_context!, result);
    }
    return await Navigator.maybePop(_context!, result);
  }

  /// Pop the current route off and push a named route in its place.
  void popAndPushNamed(String routeName, {Object? arguments}) {
    if (_delegate != null) {
      _delegate!.popAndPushNamed(_context!, routeName, arguments: arguments);
      return;
    }
    Navigator.popAndPushNamed(_context!, routeName, arguments: arguments);
  }

  /// Pops until a route with the given name.
  void popUntil(String routeName) {
    bool predicateFunc(Route<dynamic> route) => route.settings.name == routeName;
    if (_delegate != null) {
      _delegate!.popUntil(_context!, predicateFunc);
      return;
    }
    Navigator.popUntil(_context!, predicateFunc);
  }

  /// Push the route with the given name and remove routes until the predicate returns true.
  /// Original API - kept for backward compatibility
  void pushNamedAndRemoveUntil(Object? state, String newRouteName, String untilRouteName) {
    pushNamedAndRemoveUntilRoute(newRouteName, untilRouteName, arguments: state);
  }

  /// Push the route with the given name and remove routes until named route is reached.
  /// This is a more Flutter-like API
  void pushNamedAndRemoveUntilRoute(String newRouteName, String untilRouteName, {Object? arguments}) {
    bool predicateFunc(Route<dynamic> route) => route.settings.name == untilRouteName;
    if (_delegate != null) {
      _delegate!.pushNamedAndRemoveUntil(
        _context!,
        newRouteName,
        predicateFunc,
        arguments: arguments
      );
      return;
    }
    Navigator.pushNamedAndRemoveUntil(
      _context!,
      newRouteName,
      predicateFunc,
      arguments: arguments
    );
  }

  @override
  dynamic invoke(String method, Object? params, InvokeModuleCallback callback) {
    // Handle the case where params might be null
    List<dynamic> paramsList = [];
    if (params != null) {
      paramsList = params as List<dynamic>;
    }

    if (_context == null) {
      throw FlutterError('Could not invoke HybridHistory API when flutter context was not attached');
    }

    switch (method) {
      case 'state':
        Map<String, dynamic>? initialState = moduleManager!.controller.initialState;

        if (_delegate != null) {
          return _delegate!.state(_context!, initialState);
        }

        var route = ModalRoute.of(_context!);

        if (route?.settings.arguments != null) {
          return jsonEncode(route!.settings.arguments);
        } else if (route?.settings.name == null && initialState != null) {
          return jsonEncode(initialState);
        }
        return '{}';

      // Original API methods - for backward compatibility
      case 'back':
        back();
        break;
      case 'pushState':
        if (paramsList.length >= 2) {
          pushState(paramsList[0], paramsList[1]);
        }
        break;
      case 'replaceState':
        if (paramsList.length >= 2) {
          replaceState(paramsList[0], paramsList[1]);
        }
        break;
      case 'restorablePopAndPushState':
        if (paramsList.length >= 2) {
          return restorablePopAndPushState(paramsList[0], paramsList[1]);
        }
        break;

      // New Flutter-like API methods
      case 'pop':
        pop(paramsList.isNotEmpty ? paramsList[0] : null);
        break;
      case 'pushNamed':
        if (paramsList.length >= 2) {
          pushNamed(paramsList[0], arguments: paramsList[1]);
        } else if (paramsList.isNotEmpty) {
          pushNamed(paramsList[0]);
        }
        break;
      case 'pushReplacementNamed':
        if (paramsList.length >= 2) {
          pushReplacementNamed(paramsList[0], arguments: paramsList[1]);
        } else if (paramsList.isNotEmpty) {
          pushReplacementNamed(paramsList[0]);
        }
        break;
      case 'restorablePopAndPushNamed':
        if (paramsList.length >= 2) {
          return restorablePopAndPushNamed(paramsList[0], arguments: paramsList[1]);
        } else if (paramsList.isNotEmpty) {
          return restorablePopAndPushNamed(paramsList[0]);
        }
        break;

      // Other methods
      case 'path':
        return path();
      case 'canPop':
        return canPop().toString();
      case 'maybePop':
        // We need to handle this specially because it returns a Future
        maybePop(paramsList.isNotEmpty ? paramsList[0] : null).then((value) {
          callback(data: value.toString());
        });
        return EMPTY_STRING;
      case 'popAndPushNamed':
        if (paramsList.length >= 2) {
          popAndPushNamed(paramsList[0], arguments: paramsList[1]);
        } else if (paramsList.isNotEmpty) {
          popAndPushNamed(paramsList[0]);
        }
        break;
      case 'popUntil':
        if (paramsList.isNotEmpty) {
          popUntil(paramsList[0]);
        }
        break;
      case 'pushNamedAndRemoveUntil':
        if (paramsList.length == 3) {
          // Old style with 3 params
          pushNamedAndRemoveUntil(paramsList[0], paramsList[1], paramsList[2]);
        } else if (paramsList.length >= 2) {
          // New style with 2+ params
          pushNamedAndRemoveUntilRoute(paramsList[0], paramsList[1],
            arguments: paramsList.length > 2 ? paramsList[2] : null);
        }
        break;
    }
    return EMPTY_STRING;
  }

  @override
  void dispose() {}
}
